beginner_source/profiler_tutorial.py

```diff
@@ -0,0 +1,205 @@
+######################################################################
+# Performance debugging using Profiler
+# ~~~~~~~~~~~~~~~~~
+#
+# Profiler can be useful to identify performance bottlenecks in your
+# models. In this example, we build a custom module that performs two
+# sub-tasks:
+#
+# - a linear transformation on the input, and
+# - use the transformation result to get indices on a mask tensor.
+#
+# We wrap the code for each sub-task in separate labelled context managers using
+# ``profiler.record_function("label")``. In the profiler output, the
+# aggregate performance metrics of all operations in the sub-task will
+# show up under its corresponding label.
+#
+#
+# Note that this incurs some overhead, and is best used only while
+# investigating the trace. Remove it if you are benchmarking runtimes.
+#
+
+class CustomLinear(torch.nn.Linear):
+    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
+        super().__init__(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        # with profiler.record_function("LINEAR PASS"):
+            out = super().forward(input)
+
+        # with profiler.record_function("MASK INDICES"):
+            threshold = out.sum(axis=1).mean().item()
+            hi_idx = np.argwhere(mask.cpu().numpy() > threshold) # fix: use torch.nonzero()
+
+            return out, hi_idx
+
+
+######################################################################
+# Profile the forward pass
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# We initialize random input and mask tensors, and the ``CustomLinear`` module.
+# Before we run the profiler, we warm-up CUDA to ensure accurate
+# performance benchmarking. We wrap the forward pass of our module in the
+# ``profiler.profile`` context manager. The ``with_stack=True`` parameter appends the
+# file and line number of the operation in the trace.
+#
+
+model = CustomLinear(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.double).cuda()
+
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+
+######################################################################
+# Pretty print profiler results
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Finally, we print the profiler results. ``profiler.key_averages``
+# aggregates the results either by input shapes or stack trace events.
+# Grouping by input shapes is useful to identify which tensor shapes
+# contribute to runtime.
+#
+# Here, we use ``group_by_stack_n=5`` which aggregates runtimes by the
+# operation and its traceback (truncated to the most recent 5 events), and
+# display the events in the order they are registered. The table can also
+# be sorted by passing a ``sort_by`` argument (refer to the
+# `docs <https://pytorch.org/docs/stable/autograd.html#profiler>`__ for
+# valid sorting keys).
+#
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
```

Comment:
we should also include the output of the profiler

Reply:
The output doesn't show up in the netlify preview, but will build and show when deployed